File: src/tracking.rs
use std::fs;
use std::io::{Read, Write, Seek, SeekFrom};
use std::fs::File;
use std::time::{SystemTime, UNIX_EPOCH};

pub const TRACKING_DIR: &str = "/var/lib/Qube";
pub const CONTAINER_LIST_FILE: &str = "/var/lib/Qube/containers.txt";

#[derive(Debug)]
pub struct ContainerEntry {
    pub name: String,
    pub pid: i32,
    pub dir: String,
    pub command: Vec<String>,
    pub timestamp: u64,
}

fn current_timestamp() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs()
}

/// Format: name|pid|dir|command|timestamp
pub fn track_container_named(n: &str, p: i32, d: &str, c: Vec<String>) {
    fs::create_dir_all(TRACKING_DIR).ok();
    let timestamp = current_timestamp();
    let s = c.join("\t");
    let line = format!("{}|{}|{}|{}|{}", n, p, d, s, timestamp);

    let mut f = fs::OpenOptions::new()
        .create(true)
        .append(true)
        .open(CONTAINER_LIST_FILE)
        .expect("Failed to open container tracking file");

    if let Ok(metadata) = fs::metadata(CONTAINER_LIST_FILE) {
        if metadata.len() > 0 {
            let mut file = File::open(CONTAINER_LIST_FILE).unwrap();
            file.seek(SeekFrom::End(-1)).unwrap();
            let mut buf = [0u8; 1];
            file.read_exact(&mut buf).unwrap();
            if buf[0] != b'\n' {
                f.write_all(b"\n").unwrap();
            }
        }
    }

    writeln!(f, "{}", line).unwrap();
}

pub fn update_container_pid(name: &str, new_pid: i32, new_dir: &str, new_cmd: &[String]) {
    let mut found = false;
    let new_line = format!("{}|{}|{}|{}|{}", name, new_pid, new_dir, new_cmd.join("\t"), current_timestamp());
    if let Ok(c) = fs::read_to_string(CONTAINER_LIST_FILE) {
        let mut new_lines = Vec::new();
        for l in c.lines() {
            let parts: Vec<&str> = l.splitn(5, '|').collect();
            if parts.len() < 4 {
                new_lines.push(l.to_string());
                continue;
            }
            if parts[0] == name {
                new_lines.push(new_line.clone());
                found = true;
            } else {
                new_lines.push(l.to_string());
            }
        }
        let joined = new_lines.join("\n");
        fs::write(CONTAINER_LIST_FILE, joined).unwrap();
    }
    if !found {
        track_container_named(name, new_pid, new_dir, new_cmd.to_vec());
    }
}

pub fn remove_container_from_tracking(pid: i32) {
    if let Ok(c) = fs::read_to_string(CONTAINER_LIST_FILE) {
        let mut nc = Vec::new();
        for l in c.lines() {
            let parts: Vec<&str> = l.splitn(5, '|').collect();
            if parts.len() < 4 {
                continue;
            }
            if let Ok(x) = parts[1].parse::<i32>() {
                if x == pid {
                    continue;
                }
            }
            nc.push(l.to_string());
        }
        let joined = nc.join("\n");
        fs::write(CONTAINER_LIST_FILE, joined).unwrap();
    }
}

pub fn remove_container_from_tracking_by_name(name: &str) {
    if let Ok(c) = fs::read_to_string(CONTAINER_LIST_FILE) {
        let new_lines: Vec<String> = c
            .lines()
            .filter(|l| {
                let parts: Vec<&str> = l.splitn(5, '|').collect();
                if parts.len() < 4 {
                    return true;
                }
                parts[0] != name
            })
            .map(String::from)
            .collect();
        let joined = new_lines.join("\n");
        fs::write(CONTAINER_LIST_FILE, joined).unwrap();
    }
}

pub fn get_all_tracked_entries() -> Vec<ContainerEntry> {
    let mut v = Vec::new();
    if let Ok(c) = fs::read_to_string(CONTAINER_LIST_FILE) {
        for l in c.lines() {
            let parts: Vec<&str> = l.splitn(5, '|').collect();
            if parts.len() < 4 {
                continue;
            }
            let name = parts[0].to_string();
            let pid = parts[1].parse::<i32>().unwrap_or(-1);
            let dir = parts[2].to_string();
            let cmd_str = parts[3];
            let cmd_parts: Vec<String> = cmd_str.split('\t').map(|s| s.to_string()).collect();
            let timestamp = if parts.len() >= 5 {
                parts[4].parse::<u64>().unwrap_or(0)
            } else {
                0
            };
            let e = ContainerEntry { name, pid, dir, command: cmd_parts, timestamp };
            v.push(e);
        }
    }
    v
}

pub fn get_process_uptime(pid: i32) -> Result<u64, std::io::Error> {
    let p = format!("/proc/{}/stat", pid);
    let mut b = String::new();
    File::open(&p)?.read_to_string(&mut b)?;
    let f: Vec<&str> = b.split_whitespace().collect();
    if f.len() <= 21 {
        return Err(std::io::Error::new(std::io::ErrorKind::Other, "Failed to parse stat"));
    }
    let st: f64 = f[21].parse().unwrap_or(0.0);
    let mut us = String::new();
    File::open("/proc/uptime")?.read_to_string(&mut us)?;
    let su: f64 = us.split_whitespace().next().unwrap_or("0").parse().unwrap_or(0.0);
    let hz = 100.0;
    let ss = st / hz;
    let pu = su - ss;
    Ok(pu.max(0.0) as u64)
}

File: src/core/tracking/tracking.go
package tracking

import (
	"bufio"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"

	"github.com/Voyrox/Qube/src/config"
)

type ContainerEntry struct {
	Name      string
	PID       int
	Dir       string
	Command   []string
	Timestamp uint64
	Image     string
	Ports     string
	Isolated  bool
	Volumes   [][2]string
	EnvVars   []string
}

func currentTimestamp() uint64 {
	return uint64(time.Now().Unix())
}

func encodeTrackingField(v any) (string, error) {
	if v == nil {
		return "", nil
	}

	raw, err := json.Marshal(v)
	if err != nil {
		return "", err
	}

	return base64.StdEncoding.EncodeToString(raw), nil
}

func decodeVolumesField(raw string) [][2]string {
	if raw == "" {
		return [][2]string{}
	}

	decoded, err := base64.StdEncoding.DecodeString(raw)
	if err != nil {
		return [][2]string{}
	}

	var volumes [][2]string
	if err := json.Unmarshal(decoded, &volumes); err != nil {
		return [][2]string{}
	}

	return volumes
}

func decodeEnvVarsField(raw string) []string {
	if raw == "" {
		return []string{}
	}

	decoded, err := base64.StdEncoding.DecodeString(raw)
	if err != nil {
		return []string{}
	}

	var envVars []string
	if err := json.Unmarshal(decoded, &envVars); err != nil {
		return []string{}
	}

	return envVars
}

func TrackContainerNamed(name string, pid int, dir string, cmd []string, image string, ports string, isolated bool, volumes [][2]string, envVars []string) error {
	if err := os.MkdirAll(config.TrackingDir, 0755); err != nil {
		return err
	}

	timestamp := currentTimestamp()
	cmdStr := strings.Join(cmd, "\t")
	encodedVolumes, err := encodeTrackingField(volumes)
	if err != nil {
		return err
	}
	encodedEnvVars, err := encodeTrackingField(envVars)
	if err != nil {
		return err
	}
	line := fmt.Sprintf("%s|%d|%s|%s|%d|%s|%s|%t|%s|%s\n", name, pid, dir, cmdStr, timestamp, image, ports, isolated, encodedVolumes, encodedEnvVars)

	f, err := os.OpenFile(config.ContainerListFile, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0644)
	if err != nil {
		return err
	}
	defer f.Close()

	info, err := os.Stat(config.ContainerListFile)
	if err == nil && info.Size() > 0 {
		content, _ := os.ReadFile(config.ContainerListFile)
		if len(content) > 0 && content[len(content)-1] != '\n' {
			f.WriteString("\n")
		}
	}

	_, err = f.WriteString(line)
	return err
}

func UpdateContainerPID(name string, newPID int, newDir string, newCmd []string, image string, ports string, isolated bool, volumes [][2]string, envVars []string) error {
	found := false
	encodedVolumes, err := encodeTrackingField(volumes)
	if err != nil {
		return err
	}
	encodedEnvVars, err := encodeTrackingField(envVars)
	if err != nil {
		return err
	}
	newLine := fmt.Sprintf("%s|%d|%s|%s|%d|%s|%s|%t|%s|%s", name, newPID, newDir, strings.Join(newCmd, "\t"), currentTimestamp(), image, ports, isolated, encodedVolumes, encodedEnvVars)

	content, err := os.ReadFile(config.ContainerListFile)
	if err != nil {
		if os.IsNotExist(err) {
			return TrackContainerNamed(name, newPID, newDir, newCmd, image, ports, isolated, volumes, envVars)
		}
		return err
	}

	lines := strings.Split(string(content), "\n")
	var newLines []string

	for _, line := range lines {
		if line == "" {
			continue
		}
		parts := strings.Split(line, "|")
		if len(parts) < 4 {
			newLines = append(newLines, line)
			continue
		}
		if parts[0] == name {
			newLines = append(newLines, newLine)
			found = true
		} else {
			newLines = append(newLines, line)
		}
	}

	if !found {
		return TrackContainerNamed(name, newPID, newDir, newCmd, image, ports, isolated, volumes, envVars)
	}

	return os.WriteFile(config.ContainerListFile, []byte(strings.Join(newLines, "\n")+"\n"), 0644)
}

func RemoveContainerFromTracking(pid int) error {
	content, err := os.ReadFile(config.ContainerListFile)
	if err != nil {
		return err
	}

	lines := strings.Split(string(content), "\n")
	var newLines []string

	for _, line := range lines {
		if line == "" {
			continue
		}
		parts := strings.Split(line, "|")
		if len(parts) < 4 {
			continue
		}
		linePID, err := strconv.Atoi(parts[1])
		if err == nil && linePID == pid {
			continue
		}
		newLines = append(newLines, line)
	}

	return os.WriteFile(config.ContainerListFile, []byte(strings.Join(newLines, "\n")+"\n"), 0644)
}

func RemoveContainerFromTrackingByName(name string) error {
	content, err := os.ReadFile(config.ContainerListFile)
	if err != nil {
		return err
	}

	lines := strings.Split(string(content), "\n")
	var newLines []string

	for _, line := range lines {
		if line == "" {
			continue
		}
		parts := strings.Split(line, "|")
		if len(parts) < 4 {
			newLines = append(newLines, line)
			continue
		}
		if parts[0] != name {
			newLines = append(newLines, line)
		}
	}

	return os.WriteFile(config.ContainerListFile, []byte(strings.Join(newLines, "\n")+"\n"), 0644)
}

func GetAllTrackedEntries() []ContainerEntry {
	var entries []ContainerEntry

	file, err := os.Open(config.ContainerListFile)
	if err != nil {
		return entries
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		if line == "" {
			continue
		}
		parts := strings.Split(line, "|")
		if len(parts) < 8 {
			continue
		}

		pid, _ := strconv.Atoi(parts[1])
		timestamp, _ := strconv.ParseUint(parts[4], 10, 64)
		cmdParts := strings.Split(parts[3], "\t")
		isolated := strings.TrimSpace(parts[7]) == "true"

		entry := ContainerEntry{
			Name:      parts[0],
			PID:       pid,
			Dir:       parts[2],
			Command:   cmdParts,
			Timestamp: timestamp,
			Image:     parts[5],
			Ports:     parts[6],
			Isolated:  isolated,
			Volumes: decodeVolumesField(func() string {
				if len(parts) > 8 {
					return parts[8]
				}
				return ""
			}()),
			EnvVars: decodeEnvVarsField(func() string {
				if len(parts) > 9 {
					return parts[9]
				}
				return ""
			}()),
		}
		entries = append(entries, entry)
	}

	return entries
}

func GetProcessUptime(pid int) (uint64, error) {
	statPath := fmt.Sprintf("/proc/%d/stat", pid)
	content, err := os.ReadFile(statPath)
	if err != nil {
		return 0, err
	}

	fields := strings.Fields(string(content))
	if len(fields) <= 21 {
		return 0, fmt.Errorf("invalid stat file format")
	}

	startTime, err := strconv.ParseUint(fields[21], 10, 64)
	if err != nil {
		return 0, err
	}

	uptimeContent, err := os.ReadFile("/proc/uptime")
	if err != nil {
		return 0, err
	}

	uptimeFields := strings.Fields(string(uptimeContent))
	if len(uptimeFields) < 1 {
		return 0, fmt.Errorf("invalid uptime file")
	}

	systemUptime, err := strconv.ParseFloat(uptimeFields[0], 64)
	if err != nil {
		return 0, err
	}

	clkTck := uint64(100)
	processStartSecs := startTime / clkTck
	uptime := uint64(systemUptime) - processStartSecs

	return uptime, nil
}
